File: pharmacy_management_system/server/routes/customers.ts
import { Router } from "express";
import { pool, ensureConnection } from "../db.js";

const router = Router();

// Get all customers
router.get("/", async (req, res) => {
  try {
    await ensureConnection();
    const { page = 1, limit = 50, search } = req.query;
    
    let query = "SELECT * FROM customers WHERE 1=1";
    const params: any[] = [];
    let paramCount = 1;
    
    if (search) {
      query += ` AND (name ILIKE $${paramCount} OR phone ILIKE $${paramCount} OR email ILIKE $${paramCount})`;
      params.push(`%${search}%`);
      paramCount++;
    }
    
    const offset = (Number(page) - 1) * Number(limit);
    query += ` ORDER BY created_at DESC LIMIT $${paramCount} OFFSET $${paramCount + 1}`;
    params.push(Number(limit), offset);
    
    const result = await pool.query(query, params);
    
    // Get total count
    let countQuery = "SELECT COUNT(*) FROM customers WHERE 1=1";
    const countParams: any[] = [];
    if (search) {
      countQuery += " AND (name ILIKE $1 OR phone ILIKE $1 OR email ILIKE $1)";
      countParams.push(`%${search}%`);
    }
    const countResult = await pool.query(countQuery, countParams);
    const total = parseInt(countResult.rows[0].count);
    
    res.json({
      data: result.rows,
      pagination: {
        page: Number(page),
        limit: Number(limit),
        total,
        totalPages: Math.ceil(total / Number(limit)),
      },
    });
  } catch (error: any) {
    console.error("Error fetching customers:", error);
    res.status(500).json({ error: "Failed to fetch customers", details: error.message });
  }
});

// Get single customer with purchase history
router.get("/:id", async (req, res) => {
  try {
    await ensureConnection();
    const { id } = req.params;
    
    const customerResult = await pool.query(
      "SELECT * FROM customers WHERE id = $1",
      [id]
    );
    
    if (customerResult.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    
    const salesResult = await pool.query(
      `SELECT s.*, 
              (SELECT COUNT(*) FROM sales_items WHERE sale_id = s.id) as item_count
       FROM sales s
       WHERE s.customer_id = $1
       ORDER BY s.created_at DESC
       LIMIT 50`,
      [id]
    );
    
    const prescriptionsResult = await pool.query(
      "SELECT * FROM prescriptions WHERE customer_id = $1 ORDER BY created_at DESC",
      [id]
    );
    
    res.json({
      ...customerResult.rows[0],
      purchaseHistory: salesResult.rows,
      prescriptions: prescriptionsResult.rows,
    });
  } catch (error: any) {
    console.error("Error fetching customer:", error);
    res.status(500).json({ error: "Failed to fetch customer", details: error.message });
  }
});

// Create new customer
router.post("/", async (req, res) => {
  try {
    await ensureConnection();
    const {
      name,
      email,
      phone,
      address,
      date_of_birth,
      gender,
    } = req.body;
    
    if (!name) {
      return res.status(400).json({ error: "Customer name is required" });
    }
    
    const result = await pool.query(
      `INSERT INTO customers (name, email, phone, address, date_of_birth, gender)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [name, email || null, phone || null, address || null, date_of_birth || null, gender || null]
    );
    
    res.status(201).json(result.rows[0]);
  } catch (error: any) {
    console.error("Error creating customer:", error);
    if (error.code === "23505") {
      return res.status(400).json({ error: "Customer with this email or phone already exists" });
    }
    res.status(500).json({ error: "Failed to create customer", details: error.message });
  }
});

// Update customer
router.put("/:id", async (req, res) => {
  try {
    await ensureConnection();
    const { id } = req.params;
    const updates = req.body;
    
    const allowedFields = ["name", "email", "phone", "address", "date_of_birth", "gender"];
    const updateFields: string[] = [];
    const params: any[] = [];
    let paramCount = 1;
    
    Object.keys(updates).forEach((key) => {
      if (allowedFields.includes(key)) {
        updateFields.push(`${key} = $${paramCount}`);
        params.push(updates[key]);
        paramCount++;
      }
    });
    
    if (updateFields.length === 0) {
      return res.status(400).json({ error: "No valid fields to update" });
    }
    
    updateFields.push(`updated_at = CURRENT_TIMESTAMP`);
    params.push(id);
    
    const result = await pool.query(
      `UPDATE customers SET ${updateFields.join(", ")} WHERE id = $${paramCount} RETURNING *`,
      params
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    
    res.json(result.rows[0]);
  } catch (error: any) {
    console.error("Error updating customer:", error);
    res.status(500).json({ error: "Failed to update customer", details: error.message });
  }
});

// Delete customer
router.delete("/:id", async (req, res) => {
  try {
    await ensureConnection();
    const { id } = req.params;
    const result = await pool.query("DELETE FROM customers WHERE id = $1 RETURNING *", [id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    
    res.json({ message: "Customer deleted successfully" });
  } catch (error: any) {
    console.error("Error deleting customer:", error);
    res.status(500).json({ error: "Failed to delete customer", details: error.message });
  }
});

export default router;

File: pharmacy_management_system/python_backend/routes/customers.py
"""
Customers API Routes
Handles customer management and purchase history
"""
from flask import Blueprint, request, jsonify
from db import execute_query
import logging

logger = logging.getLogger(__name__)
customers_bp = Blueprint('customers', __name__)

@customers_bp.route('/', methods=['GET'])
def get_customers():
    """Get all customers with pagination and search"""
    try:
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 50))
        search = request.args.get('search', '')
        
        query = "SELECT * FROM customers WHERE 1=1"
        count_query = "SELECT COUNT(*) as count FROM customers WHERE 1=1"
        params = []
        
        if search:
            query += " AND (name ILIKE %s OR phone ILIKE %s OR email ILIKE %s)"
            count_query += " AND (name ILIKE %s OR phone ILIKE %s OR email ILIKE %s)"
            search_param = f'%{search}%'
            params.extend([search_param, search_param, search_param])
        
        # Get total count
        total_result = execute_query(count_query, tuple(params) if params else None, fetch_one=True)
        total = total_result['count'] if total_result else 0
        
        # Add pagination
        offset = (page - 1) * limit
        query += " ORDER BY created_at DESC LIMIT %s OFFSET %s"
        params.extend([limit, offset])
        
        customers = execute_query(query, tuple(params) if params else None)
        
        return jsonify({
            'data': customers,
            'pagination': {
                'page': page,
                'limit': limit,
                'total': total,
                'totalPages': (total + limit - 1) // limit
            }
        }), 200
    except Exception as e:
        logger.error(f"Error fetching customers: {e}")
        return jsonify({'error': 'Failed to fetch customers', 'details': str(e)}), 500

@customers_bp.route('/<int:customer_id>', methods=['GET'])
def get_customer(customer_id):
    """Get single customer with purchase history"""
    try:
        # Get customer
        customer_query = "SELECT * FROM customers WHERE id = %s"
        customer = execute_query(customer_query, (customer_id,), fetch_one=True)
        
        if not customer:
            return jsonify({'error': 'Customer not found'}), 404
        
        # Get purchase history
        sales_query = """
            SELECT s.*, 
                   (SELECT COUNT(*) FROM sales_items WHERE sale_id = s.id) as item_count
            FROM sales s
            WHERE s.customer_id = %s
            ORDER BY s.created_at DESC
            LIMIT 50
        """
        purchase_history = execute_query(sales_query, (customer_id,))
        
        # Get prescriptions
        prescriptions_query = """
            SELECT * FROM prescriptions 
            WHERE customer_id = %s 
            ORDER BY created_at DESC
        """
        prescriptions = execute_query(prescriptions_query, (customer_id,))
        
        result = dict(customer)
        result['purchaseHistory'] = purchase_history
        result['prescriptions'] = prescriptions
        
        return jsonify(result), 200
    except Exception as e:
        logger.error(f"Error fetching customer: {e}")
        return jsonify({'error': 'Failed to fetch customer', 'details': str(e)}), 500

@customers_bp.route('/', methods=['POST'])
def create_customer():
    """Create new customer"""
    try:
        data = request.get_json()
        
        if not data.get('name'):
            return jsonify({'error': 'Customer name is required'}), 400
        
        query = """
            INSERT INTO customers (name, email, phone, address, date_of_birth, gender)
            VALUES (%s, %s, %s, %s, %s, %s)
            RETURNING *
        """
        params = (
            data.get('name'),
            data.get('email'),
            data.get('phone'),
            data.get('address'),
            data.get('date_of_birth'),
            data.get('gender')
        )
        
        customer = execute_query(query, params, fetch_one=True)
        return jsonify(customer), 201
    except Exception as e:
        logger.error(f"Error creating customer: {e}")
        if 'duplicate key' in str(e).lower():
            return jsonify({'error': 'Customer with this email or phone already exists'}), 400
        return jsonify({'error': 'Failed to create customer', 'details': str(e)}), 500

@customers_bp.route('/<int:customer_id>', methods=['PUT'])
def update_customer(customer_id):
    """Update customer"""
    try:
        data = request.get_json()
        
        allowed_fields = ['name', 'email', 'phone', 'address', 'date_of_birth', 'gender']
        update_fields = []
        params = []
        
        for field in allowed_fields:
            if field in data:
                update_fields.append(f"{field} = %s")
                params.append(data[field])
        
        if not update_fields:
            return jsonify({'error': 'No valid fields to update'}), 400
        
        update_fields.append("updated_at = CURRENT_TIMESTAMP")
        params.append(customer_id)
        
        query = f"UPDATE customers SET {', '.join(update_fields)} WHERE id = %s RETURNING *"
        customer = execute_query(query, tuple(params), fetch_one=True)
        
        if not customer:
            return jsonify({'error': 'Customer not found'}), 404
        
        return jsonify(customer), 200
    except Exception as e:
        logger.error(f"Error updating customer: {e}")
        return jsonify({'error': 'Failed to update customer', 'details': str(e)}), 500

@customers_bp.route('/<int:customer_id>', methods=['DELETE'])
def delete_customer(customer_id):
    """Delete customer"""
    try:
        query = "DELETE FROM customers WHERE id = %s RETURNING *"
        customer = execute_query(query, (customer_id,), fetch_one=True)
        
        if not customer:
            return jsonify({'error': 'Customer not found'}), 404
        
        return jsonify({'message': 'Customer deleted successfully'}), 200
    except Exception as e:
        logger.error(f"Error deleting customer: {e}")
        return jsonify({'error': 'Failed to delete customer', 'details': str(e)}), 500
